cmd/controller/app/controller.go

```diff
@@ -63,89 +66,145 @@ const controllerAgentName = "cert-manager"
 //and following discussion: https://github.com/kubernetes-sigs/controller-runtime/pull/88#issuecomment-408500629
 const resyncPeriod = 10 * time.Hour
 
-func Run(opts *options.ControllerOptions, stopCh <-chan struct{}) {
+func Run(opts *options.ControllerOptions, stopCh <-chan struct{}) error {
 	rootCtx := cmdutil.ContextWithStopCh(context.Background(), stopCh)
+	rootCtx, cancelContext := context.WithCancel(rootCtx)
+	defer cancelContext()
+	g, rootCtx := errgroup.WithContext(rootCtx)
 	rootCtx = logf.NewContext(rootCtx, nil, "controller")
 	log := logf.FromContext(rootCtx)
 
-	ctx, kubeCfg, err := buildControllerContext(rootCtx, stopCh, opts)
+	ctx, kubeCfg, err := buildControllerContext(rootCtx, opts)
 	if err != nil {
 		log.Error(err, "error building controller context", "options", opts)
-		os.Exit(1)
+		return err
 	}
 
 	enabledControllers := opts.EnabledControllers()
 	log.Info(fmt.Sprintf("enabled controllers: %s", enabledControllers.List()))
 
-	metricsServer, err := ctx.Metrics.Start(opts.MetricsListenAddress, opts.EnablePprof)
+	ln, err := net.Listen("tcp", opts.MetricsListenAddress)
 	if err != nil {
 		log.Error(err, "failed to listen on prometheus address", "address", opts.MetricsListenAddress)
-		os.Exit(1)
+		return err
 	}
+	server := ctx.Metrics.NewServer(ln, opts.EnablePprof)
 
-	var wg sync.WaitGroup
-	run := func(_ context.Context) {
-		for n, fn := range controller.Known() {
-			log := log.WithValues("controller", n)
+	g.Go(func() error {
+		<-rootCtx.Done()
+		// allow a timeout for graceful shutdown
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
 
-			// only run a controller if it's been enabled
-			if !enabledControllers.Has(n) {
-				log.V(logf.InfoLevel).Info("not starting controller as it's disabled")
-				continue
+		if err := server.Shutdown(ctx); err != nil {
+			return err
+		}
+		return nil
+	})
+	g.Go(func() error {
+		log.WithValues("address", ln.Addr()).V(logf.InfoLevel).Info("listening for connections on")
```

Comment:
I can't see this message when the controller starts. Can you?

And I suggest the following message instead.

```suggestion
		log.V(logf.InfoLevel).Info("starting metrics server", "address", ln.Addr())
```

Reply:
I changed  it, `bazel run //cmd/controller:controller -- --kubeconfig ~/.kube/config -v 3` shows that info msg on my machine.